text/0000-socket-timeouts.md

```diff
@@ -0,0 +1,142 @@
+- Feature Name: socket_timeouts
+- Start Date: 2015-04-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add sockopt-style timeouts to `std::net` types.
+
+# Motivation
+
+Currently, operations on various socket types in `std::net` block
+indefinitely (i.e., until the connection is closed or data is
+transferred). But there are many contexts in which timing out a
+blocking call is important.
+
+The [goal of the current IO system][io-reform] is to gradually expose
+cross-platform, blocking APIs for IO, especially APIs that directly
+correspond to the underlying system APIs. Sockets are widely available
+with nearly identical system APIs across the platforms Rust targets,
+and this includes support for timeouts via [sockopts][sockopt].
+
+So timeouts are well-motivated and well-suited to `std::net`.
+
+# Detailed design
+
+The proposal is to *directly expose* the timeout functionality
+provided by [`setsockopt`][sockopt], in much the same way we currently
+expose functionality like `set_nodelay`:
+
+```rust
+impl TcpStream {
+    pub fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
+    pub fn read_timeout(&self) -> Option<Duration>;
```

Comment:
The way it would be represented in other languages matches `Result<Option<Duration>>`. For instance, Python would be:

```python
try:
  timeout = so.gettimeout()
  if timeout is not None:
    # whatever
except OSError, e:
  # handle error
```